packages/material-ui/src/TablePagination/TablePagination.test.js

```diff
@@ -425,4 +425,28 @@ describe('<TablePagination />', () => {
       expect(container).to.include.text('1-25 of 25');
     });
   });
+
+  describe('duplicated keys', () => {
+    it('should not raise a warning due to duplicated keys', () => {
+      render(
+        <table>
+          <TableFooter>
+            <TableRow>
+              <TablePagination
+                rowsPerPageOptions={[5, 10, { label: 'All', value: 10 }]}
```

Comment:
I honestly don't really understand what we're trying to fix here. Why would you add two options with the same value? It seems to me that the existing behavior is correct but the error message is not helpful. Ideally we'd just warn that duplicate values are provided.

Reply:
Quite simply, `value` can be dynamic here, which was the case in our application which resulted in the duplicated key in the first place.
To be exact, we use a table which contains data, and when you add data to it, it appends rows to it. The `rowsPerPageOptions` are fixed (in our case it is something like `[5, 10, 20, { label: 'All', value: data.length }]`) so if it so happens, that `data.length` is 10, we get a duplicate key error.

Sure we could also implement some logic to dynamically compute the `rowsPerPageOptions` but honestly at this point I think it would be over-engineering, especially considering this approach works fine and it makes sure even if you choose `All` and add new data to the table, that even if it goes beyond `10` you still see all options.

Considering the use case is valid in my opinion, I don't think it would make more sense to display a different error message instead. In the end it's not about any functionality we're talking of here, it's just about keys, which React internally uses to track changes.

Reply:
> Sure we could also implement some logic to dynamically compute the rowsPerPageOptions but honestly at this point I think it would be over-engineering

Instead you're degrading the library for everybody else. We agree that displaying multiple options for the same value is bad. You just think that you shouldn't fix this to make your life easier. I understand the problem space but it seems to me that using `label` is strictly better than `index` and `value` while `index` is worse than `value` in most cases.

Reply:
I think there must be a misunderstanding. What I meant with:

> Sure we could also implement some logic to dynamically compute the rowsPerPageOptions but honestly at this point I think it would be over-engineering

Was that if material-ui chose not to support the use case I mentioned and would impose the restriction that all of the values in `rowsPerPageOptions` must be unique, I could **technically** implement some logic to dynamically compute the `rowsPerPageOptions`, which would result in guaranteed unique keys (meaning it wouldn't be technically impossible for me to use `TablePagination` at all). However,  this would result in a **different** and not the desired user experience, which is why I think doing so would be over-engineered (working around a problem that shouldn't exist in the first place in my opinion). I still believe the use case is valid and simple: we want to consistently and always show the user the options `5, 10, 20, All` no matter if there are 5, 7, 10 20 or 30 rows in the table. Do you think this doesn't make sense?

What I do agree on is that if you have a table with a fixed amount of rows without the possibility of adding or removing rows, it would not make sense to have `5, 10, All` `rowsPerPageOptions` if there are only 10 rows. But in practice, you will have a dynamic amount of rows and will want to be able to add or remove rows in a lot of cases. This test case is just a "snapshot" to reproduce an issue we encountered in practice and simply the most minimal way to reproduce the issue (hence why it may look a bit odd at first glance).

> We agree that displaying multiple options for the same value is bad.

No, that's not my opinion. In contrary, I think this would result in a lot of usability issues. To name a few that come to my mind:
1. Table has 11 rows, `rowsPerPageOptions={[5, 10, { label: 'All', value: 11 }]}`, (`value: 11` is bound to `data.length`) now the user deletes one row, what do you do now to avoid having two times the same value?
    1. Dynamically remove the `10`, and switch the current `rowsPerPage` to `All` for the user, now if the user adds 2 rows, he sees all rows, but would expect only to see 10.
    2. Dynamically remove the `All`. Now the user cannot change `rowsPerPage` to `All` anymore, and if they just want to see everything without having to constantly re-adjust `rowsPerPage` they need to know that when there are 11 rows they can select `All`?
2. Assuming you would want to have a set of `rowPerPageOptions`, like `5, 10, 20, 50, 100, All` on every page for consistency reasons (something I see a lot), even if there are only 40 rows for example (where it would theoretically suffice to stop at `20` and have that followed by `All`), if it so happens that `All` has `value` 50 (since there are 50 rows) what would you do?
    1. Dynamically remove the `50` option. Would be unexpected to the user, since they know from other pages they can select 50, so they assume it's a bug.
    2. Dynamically remove the `All` option. Again, The user would not be able to simply select "All" and have that adapt to always show everything. Also would be unexpected to the user, since they know from other pages they can select `All`, so they assume it's a bug.
3. User can set in their preferences how many rows they want to have displayed by default. Let's say they choose `50`. Then a table with 50 rows, same `rowsPerPageOptions` as in the last example. What do you do now?
    1. Dynamically remove the `50` option, but then what do you choose instead? `100` would not be expected, `All` would mean that when new rows are added the user will see everything and not the `50` they preset as default.
    2. Dynamically remove the `All` option. Again, The user would not be able to simply select "All"

If you know of a solution that would fulfil all of those use cases with only unique values in `rowsPerPageOptions`, I would be more than willing to take them into consideration.

> Instead you're degrading the library for everybody else. You just think that you shouldn't fix this to make your life easier.
That's not true and those are quite harsh accusations in my opinion. I can't see how you would think I'm "degrading" the library for everybody else with this PR by just changing a `key` property on a component that results in less duplicated keys? I don't have any bad or selfish intentions with this PR, be it maliciously or unintentionally. This is a code review, please stay technical.

Reply:
Agreed, yes!